src/solver/goldensectionsearch/mod.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2018-2020 argmin developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! # References:
+//!
+//! [Wikipedia](https://en.wikipedia.org/wiki/Golden-section_search)
+
+use crate::prelude::*;
+use serde::{Deserialize, Serialize};
+
+const GOLDEN_RATIO: f64 = 1.61803398874989484820;
+const G1: f64 = -1.0 + GOLDEN_RATIO;
+const G2: f64 = 1.0 - G1;
+
+/// Golden-section search
+///
+/// The golden-section search is a technique for finding an extremum (minimum or maximum) of a
+/// function inside a specified interval.
+///
+/// The method operates by successively narrowing the range of values on the specified interval,
+/// which makes it relatively slow, but very robust. The technique derives its name from the fact
+/// that the algorithm maintains the function values for four points whose three interval widths
+/// are in the ratio 2-φ:2φ-3:2-φ where φ is the golden ratio. These ratios are maintained for each
+/// iteration and are maximally efficient.
+///
+/// The `min_bound` and `max_bound` arguments define values that bracket the expected minimum. The
+/// `init_estimate` argument is the initial estimate of the minimum that is required to be larger
+/// than `min_bound` and smaller than `max_bound`.
+///
+/// # References:
+///
+/// [Wikipedia](https://en.wikipedia.org/wiki/Golden-section_search)
+#[derive(Clone, Serialize, Deserialize)]
+pub struct GoldenSectionSearch<F> {
+    g1: F,
+    g2: F,
+    min_bound: F,
+    max_bound: F,
+    init_estimate: F,
+    tolerance: F,
+
+    x0: F,
+    x1: F,
+    x2: F,
+    x3: F,
+    f1: F,
+    f2: F,
+}
+
+impl<F> GoldenSectionSearch<F>
+where
+    F: ArgminFloat,
+{
+    /// Constructor
+    pub fn new(min_bound: F, max_bound: F) -> Self {
+        GoldenSectionSearch {
+            g1: F::from(G1).unwrap(),
+            g2: F::from(G2).unwrap(),
+            min_bound,
+            max_bound,
+            init_estimate: F::zero(),
+            tolerance: F::from(0.01).unwrap(),
+            x0: min_bound,
+            x1: F::zero(),
+            x2: F::zero(),
+            x3: max_bound,
+            f1: F::zero(),
+            f2: F::zero(),
+        }
+    }
+
+    /// Set tolerance
+    pub fn tolerance(mut self, tol: F) -> Self {
+        self.tolerance = tol;
+        self
+    }
+}
+
+impl<O, F> Solver<O> for GoldenSectionSearch<F>
+where
+    O: ArgminOp<Output = F, Param = F, Float = F>,
+    F: ArgminFloat + std::cmp::PartialOrd,
```

Comment:
I think `Float` already implies `PartialOrd`: https://rust-num.github.io/num/num/trait.Float.html

Reply:
I probably added this trait before `ArgminFloat`. I removed it.